### PR TITLE
Initial pass at revisiting DatastoreQueryResults

### PR DESCRIPTION
**Pull request by jskeet**
See individual commits for details. This PR is expected to go through a fair amount of discussion and tweaking before merging.

### CODE REVIEW

File: snippets/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs [hunk 1]
```diff
@@ -155,16 +154,16 @@ public void GqlQuery()
             // DatastoreQueryResults implements IEnumerable<Entity>, but you can
```
Comment (evildour) [inline snippets/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs]: Should be IEnumerable<EntityResult>

Comment (jskeet) [inline snippets/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs]: Ack. (I'll search for AsBatches, AsResponses, `IEnumerable<Entity>` and `AsEntityResults` to try to find all stray comments.) Will probably only *do* after receiving other comments though, rather than piecemeal.

**Comment by evildour** [comment]
What is the motivation for exposing EntityResults instead of Entities by default? Is there information on the EntityResult (other than the Entity) which is common to want to get? If not, doing the extra property access or Select for each usage seems like it might be a bit tedious. Perhaps just removing the AsBatches is sufficient to simplify things. Or is the concern that the difference between Entities and EntityResults is what might be confusing?

File: src/Google.Datastore.V1Beta3/DatastoreQueryResults.cs [hunk 2]
```diff
@@ -74,22 +68,13 @@ public DatastoreQueryResults(IEnumerable<RunQueryResponse> responses)
             });
 
         /// <summary>
-        /// Returns the results of this query as a sequence of <see cref="RunQueryResponse"/> values
-        /// exactly as returned by the Datastore API. This method is only useful if the application
-        /// wishes to examine the <see cref="RunQueryResponse.Query"/> property; otherwise, use
-        /// <see cref="AsBatches"/> to obtain the sequence of batches.
+        /// This method is for advanced use cases only, where more diagnostic information is required;
```
Comment (SurferJeffAtGoogle) [inline src/Google.Datastore.V1Beta3/DatastoreQueryResults.cs]: The phrase "advanced use cases" still jumps out at me, and my ego tells me this is the function I want to use, because I'm an advanced user.

I'd rephrase:
This method is for diagnosing underlying RPC calls only.

**Comment by SurferJeffAtGoogle** [comment]
> What is the motivation for exposing EntityResults instead of Entities by default? Is there information on the EntityResult (other than the Entity) which is common to want to get? If not, doing the extra property access or Select for each usage seems like it might be a bit tedious. Perhaps just removing the AsBatches is sufficient to simplify things. Or is the concern that the difference between Entities and EntityResults is what might be confusing?

When I tried to use this API, I was overwhelmed by the 3 choices I had to retrieve entities.  I had to study each one, and read a lot of text to choose the right one, and I still chose the wrong one.  Renaming AsBatches helps a lot.  Doing the extra property access is more tedious than not, but it's fairly easy.  Note the updates to the snippets are all pretty simple.

**Comment by evildour** [comment]
> When I tried to use this API, I was overwhelmed by the 3 choices I had to retrieve entities. I had to study each one, and read a lot of text to choose the right one, and I still chose the wrong one. Renaming AsBatches helps a lot. Doing the extra property access is more tedious than not, but it's fairly easy. Note the updates to the snippets are all pretty simple.

You're right, the updates are pretty minor. Something still seems a bit off about it though. Perhaps its that what seems like the primary use case for accessing results does not come across as a primary use case for iterating the DatastoreQueryResults. Maybe it would be better to "hide" the complexity behind a helper object that gives access to rpc responses, entity results, and batches:

IEnumerable\<Entity\> entities = results;
IEnumerable\<EntityResult\> entityResults = results.RpcResult.AsEntityResults();
IEnumerable\<QueryResultBatch\> entityResults = results.RpcResult.AsBatches();
IEnumerable\<RunQueryResponse\> entityResults = results.RpcResult.AsResponses();

(I'm not suggesting that as the name for the helper, just the approach of hiding "everything else" behind it)

File: src/Google.Datastore.V1Beta3/DatastoreQueryResults.cs (hunk 2 above)
Comment (jskeet) [inline src/Google.Datastore.V1Beta3/DatastoreQueryResults.cs]: I don't want to be quite that restrictive though. Perhaps "niche use cases" would be better - ego or not, I'm sure you wouldn't argue that "displaying a page of results in a web page" is a niche use case. I'll have another look at this when I'm back from vacation, and see if any alternative wording jumps out... I'd rather have diagnostics as an example of the niche use cases than anything else, to be honest.

**Comment by jskeet** [comment]
// @pcostell for extra feedback - sorry Patrick, I should have included you before.

(Patrick is likely to have more idea of the likely usage of each of the four kinds of iteration.)

**Comment by pcostell** [comment]
Here are some use cases for the different methods:

    IEnumerable<Entity> entities = results;

I just want my entities.

    IEnumerable<EntityResult> entityResults = results.RpcResult.AsEntityResults();

I need to use cursors. This may be because I'm doing some bulk processing and need to be able to resume at any point. I also may want access to the version number so that I can do compare-and-set operations (not yet available).

    IEnumerable<QueryResultBatch> entityResults = results.RpcResult.AsBatches();

This is probably less useful as exposed. Having the QueryResultBatch is very useful for the client library, but for the user they likely only care about the last QueryResultBatch. In particular, the client library should be using the `more_results` field to track when to continue batching and use the `end_cursor` to get the next batch. For the user, they will care about the final `more_results` field to determine if they should show a "Next Page" button on their website, or continue to query for entities in an infinite scroll window. They will also care about the `end_cursor` to resume these queries.

IEnumerable<RunQueryResponse> entityResults = results.RpcResult.AsResponses();

Only the clients should really care about this. When resuming a GQL query, the client will need to pull the structured Query out of the RunQueryResponse in order to modify offsets, limits, and cursors for resuming the batch.

**Comment by jskeet** [comment]
@pcostell: Do you have any idea how often "just the entities" is useful compared with "entities and cursors"?

(The batches/responses part is definitely a niche case, although the `more_results` part does sound useful... I'd thought a web case would just use the `IEnumerable<EntityResult>`, but that could lead to an extra "Next page" button in (say) a 100-result query shown in pages of 20.)

**Comment by pcostell** [comment]
I think many times the user really just wants all their entities, my guess is that this would be the normal case (as long as the clients resume NOT_FINISHED batches as expected).

Using the `EntityResult` isn't necessarily good enough -- the web based batching you should be just running queries with limit 20 over and over again until you get `more_results == NO_MORE_RESULTS`. Also, `end_cursor` is more efficient than using the per-result cursors as it may have partial progress made after the last result.

Importantly, I think the user likely won't want to have to write the double for loop to iterate over batches and then entities within the batch. They probably want an entity iterator with side channels for getting cursors and more_results.

**Comment by jskeet** [comment]
With the implementation we've got, we update the `end_cursor` of the last `EntityResult` to make it more efficient, so it could be used... you'd just propagate the `end_cursor` on a "next" link if the page you've just fetched has as many results as you asked for. The client library makes as many requests as it should within that. Aside from the "possibly unnecessary final next page link" is there a good reason to get a batch?

@SurferJeffAtGoogle: It may be useful to link to the sample code for your web app, to make the discussion more concrete.

One extra option: have a method that eagerly fetches all the results, exposing an `IReadOnlyList<Entity>` and an `EndCursor` (which is null if there are no more results). If you're writing a web app, you're not going to return the response until you've got the whole page anyway. Maybe the laziness/streaming aspect here is just making life harder.

**Comment by jskeet** [comment]
@pcostell: Any thoughts on the idea of the extra method for the common case, where you just want all the results (up to the limit) in memory, along with an end cursor and whether or not there are more results? It wouldn't be appropriate for cases with a large or unspecified limit (and a large table) but that feels like it's probably not the use-case where you'd use it anyway.

**Comment by pcostell** [comment]
That makes sense to me. If the user is grabbing a lot of data they should probably be getting it as a stream / iterator.

**Comment by jskeet** [comment]
Closing as superceded by #343.
